.github/actions/csv-vulnerability-filter/src/processing.ts

```diff
@@ -0,0 +1,341 @@
+
+  for (const row of data) {
+    const severity = (row[severityColumn]?.trim() || 'undefined').toLowerCase();
+    switch (severity) {
+      case 'critical': counts.critical++; break;
+      case 'high': counts.high++; break;
+      case 'medium': counts.medium++; break;
+      case 'low': counts.low++; break;
+      case 'informational': counts.informational++; break;
+      default: counts.undefined++; break;
+    }
+  }
+  core.info(`Calculated counts (assuming 'Severity' column): ${JSON.stringify(counts)}`);
+  return counts;
+}
+
+export function checkCountThresholds(counts: SeverityCounts, thresholds: Thresholds): boolean {
+  let exceeded = false;
+  core.info('Checking vulnerability count thresholds...');
+
+  if (thresholds.critical >= 0 && counts.critical > thresholds.critical) {
+    core.error(`Threshold exceeded: ${counts.critical} critical vulnerabilities found (limit: ${thresholds.critical}).`);
+    exceeded = true;
+  } else if (thresholds.critical >= 0) {
+    core.info(`Critical vulnerability count (${counts.critical}) within limit (${thresholds.critical}).`);
+  }
+
+  if (thresholds.high >= 0 && counts.high > thresholds.high) {
+    core.error(`Threshold exceeded: ${counts.high} high vulnerabilities found (limit: ${thresholds.high}).`);
+    exceeded = true;
+  } else if (thresholds.high >= 0) {
+    core.info(`High vulnerability count (${counts.high}) within limit (${thresholds.high}).`);
+  }
+
+  if (thresholds.medium >= 0 && counts.medium > thresholds.medium) {
+    core.error(`Threshold exceeded: ${counts.medium} medium vulnerabilities found (limit: ${thresholds.medium}).`);
+    exceeded = true;
+  } else if (thresholds.medium >= 0) {
+    core.info(`Medium vulnerability count (${counts.medium}) within limit (${thresholds.medium}).`);
+  }
+
+  if (thresholds.low >= 0 && counts.low > thresholds.low) {
+    core.error(`Threshold exceeded: ${counts.low} low vulnerabilities found (limit: ${thresholds.low}).`);
+    exceeded = true;
+  } else if (thresholds.low >= 0) {
+    core.info(`Low vulnerability count (${counts.low}) within limit (${thresholds.low}).`);
+  }
+
+  if (!exceeded) {
+    core.info('All vulnerability counts are within defined thresholds.');
+  }
+  return exceeded;
+}
+
+export function isFixedVersionPresent(row: CsvRow): boolean {
+  const fixedVer = row['Fixed Package'];
+  if (!fixedVer) return false;
+  const trimmed = fixedVer.trim().toUpperCase();
+  return trimmed !== '' && trimmed !== 'N/A' && trimmed !== 'NONE' && !trimmed.startsWith('NO_FIX');
+}
+
+export function generateCsvString(data: CsvRow[]): string {
+  if (data.length === 0) {
+    return '';
+  }
+  try {
+    const csvString = Papa.unparse(data, {
+      header: true,
+    });
+    return csvString;
+  } catch (error) {
+    core.error(`Failed to generate CSV string: ${error instanceof Error ? error.message : String(error)}`);
```

Comment:
Once made to not b async (and moved to this file) you could do this here and a few places below:
```suggestion
    core.error(`Failed to generate CSV string: ${errorString(error)}`);
```